### PR TITLE
[react-native] Added missing type cases for ItemSeparatorComponent on FlatList

### LINKED ISSUE #18823: @types/react-native has wrong typings for FlatList> ItemSeparatorComponent
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @....
I'm currently using `FlatList` on react-native `0.47.1` 

The type definition is technically incorrect as it does not account for `React.Element<any>:`

`ItemSeparatorComponent?: React.ComponentClass<any> | null`

As a result, you will see errors like this when trying to pass in an element as a ItemSeparatorComponent:

`error TS2352: Type 'Element' cannot be converted to type 'ComponentClass<any>'. Type 'Element' provides no match for the signature 'new (props?: any, context?: any): Component<any, ComponentState>'`

From what've tried, this fixes the issue (works the same way as ListEmptyComponent, Header and Footer):
`ItemSeparatorComponent?: React.ComponentClass<any> | React.ReactElement<any> | (() => React.ReactElement<any>) | null`

@alloy , @timwangdev , @jeremistadler  would you mind checking it out?

### PR DESCRIPTION
**Pull request by mcousillas6**
Fixes #18823

- [x ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://facebook.github.io/react-native/docs/flatlist.html#itemseparatorcomponent>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Fixes #18823

### CODE REVIEW

**Review by iRoachie — APPROVED** [review]

File: types/react-native/index.d.ts [hunk 1]
```diff
@@ -3491,7 +3491,7 @@ export interface FlatListProperties<ItemT> extends VirtualizedListProperties<Ite
     /**
      * Rendered in between each item, but not at the top or bottom
      */
-    ItemSeparatorComponent?: React.ComponentClass<any> | null
+    ItemSeparatorComponent?: React.ComponentClass<any> | React.ReactElement<any> | (() => React.ReactElement<any>) | null
```
Comment (alloy) [inline types/react-native/index.d.ts]: I’m confused, [we determined](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18823#issuecomment-321985718) that this was supposed to be just: `React.ComponentClass<any> | null`, right?

Afaik it should definitely _not_ have `React.ReactElement<any>`.

**Review by alloy — CHANGES_REQUESTED** [review]

File: types/react-native/index.d.ts (hunk 1 above)
Comment (mcousillas6) [inline types/react-native/index.d.ts]: Oh, I understood that the idea was to keep the React.ComponentClass and then the other two for `ReactElement`, sorry about that misunderstanding.

From what I've tested while getting the PR ready, `React.ComponentClass` does not work for  `React.StatelesComponents`, `ItemSeparatorComponent?: React.ComponentClass<any> | null` was the one that the spec had before I opened the issue. 

I'm not entirely sure about the `ReactElement<any>` if that should not be permitted then the solution that works is `ItemSeparatorComponent?: React.ComponentClass<any> | React.StatelessComponent<any> | null`

Again, sorry about that misunderstanding.

Comment (alloy) [inline types/react-native/index.d.ts]: Oops, re-reading my last comment on the issue it seems like I suggested the wrong typing and here again, my bad. I meant `React.ComponentType<any> | null`.

`ReactElement` is afaik the output of a `React.ComponentClass` or `React.StatelessComponent`, which is not what the React Native source expects (it does `<ItemSeparatorComponent … />`)

Comment (mcousillas6) [inline types/react-native/index.d.ts]: `React.ComponentType<any> | null` Worked perfectly 👌 .

**Comment by mcousillas6** [comment]
@alloy Updated the typing to use `React.ComponentType`, worked like a charm.

**Comment by alloy** [comment]
Ace, thanks for bearing with me and your contribution 👏👏👏

**Pull request merged**